### PR TITLE
mgr/dashboard: Filter alerts based on cluster fsid and do not allow to connect clusters with version less than hub cluster in multi-cluster

### PR DESCRIPTION
**Pull request by aaSharma14**
1. Since we have a new `cluster` variable in the prometheus metrics , we need to filter the alerts based on the cluster fsid.
2. Do not allow to connect clusters with version less than 7.1 in multi-cluster

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts [hunk 1]
```diff
@@ -60,30 +62,35 @@ export class PrometheusAlertService {
...
-        alert.status.state === 'active' && alert.labels.severity === 'warning' ? ++result : result,
-      0
-    );
-    this.alerts = alerts
-      .reverse()
-      .sort((a, b) => a.labels.severity.localeCompare(b.labels.severity));
-    this.canAlertsBeNotified = true;
+    let clusterAlerts: AlertmanagerAlert[] = [];
+    this.healthService.getClusterFsid().subscribe((data: string) => {
+      const fsid = data;
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts:67]: This is again a service, filter the alert in component and pass it here to handle it.

File: src/pybind/mgr/dashboard/controllers/multi_cluster.py [hunk 2]
```diff
@@ -107,6 +107,15 @@ def check_cluster_connection(self, url, payload, username, ssl_verify, ssl_certi
...
                                          component='multi-cluster')
+            
+            multi_cluster_content = self._proxy('POST', url, 'api/multi-cluster/auth', payload=payload,
+                                  verify=ssl_verify, cert=ssl_certificate)
+            if 'status' in multi_cluster_content and multi_cluster_content['status'] == '404 Not Found':
+                raise DashboardException(msg='The ceph cluster you are attempting to connect \
+                                         to lacks the multi-cluster feature. Please ensure that \
+                                         the cluster you are connecting to is upgraded to \
+                                         IBM Ceph 7.1 to enable the multi-cluster functionality.',
+                                         code='invalid_version', component='multi-cluster')
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/controllers/multi_cluster.py:118]: The Ceph cluster you are attempting to connect \
doesn't  support Multi cluster feature. Please ensure that \
the cluster you are connecting to is upgraded to \
{Version We get from api of hub cluster} version to enable the Multi cluster feature.'

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts (hunk 1 above)
Comment (cloudbehl) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts:67]: Also there should filtering functionality in API itself as well.
Example: `api/v2/alerts/groups?filter=namespace=demo-dev&filter=alertname=DemoServiceRandomSimpleGauge`

**Comment by cloudbehl** [comment]
Merged this PR after merging this: https://github.com/ceph/ceph/pull/53679

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts [hunk 3]
```diff
@@ -23,19 +23,31 @@ export class PrometheusAlertService {
...
-  getAlerts() {
+  getAlerts(getAlertsFromFsid?: boolean) {
     this.prometheusService.ifAlertmanagerConfigured(() => {
-      this.prometheusService.getAlerts().subscribe(
-        (alerts) => this.handleAlerts(alerts),
-        (resp) => {
-          if ([404, 504].includes(resp.status)) {
-            this.prometheusService.disableAlertmanagerConfig();
+      if (getAlertsFromFsid) {
+        let params = { get_alerts_from_fsid: getAlertsFromFsid };
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts:32]:
```suggestion
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts [hunk 4]
```diff
@@ -23,19 +23,31 @@ export class PrometheusAlertService {
 
   constructor(
     private alertFormatter: PrometheusAlertFormatter,
-    private prometheusService: PrometheusService
+    private prometheusService: PrometheusService,
   ) {}
 
-  getAlerts() {
+  getAlerts(getAlertsFromFsid?: boolean) {
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts:29]:
```suggestion
  getAlerts(ClusterFilteredAlerts?: boolean) {
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts [hunk 5]
```diff
@@ -23,19 +23,31 @@ export class PrometheusAlertService {
...
     this.prometheusService.ifAlertmanagerConfigured(() => {
-      this.prometheusService.getAlerts().subscribe(
-        (alerts) => this.handleAlerts(alerts),
-        (resp) => {
-          if ([404, 504].includes(resp.status)) {
-            this.prometheusService.disableAlertmanagerConfig();
+      if (getAlertsFromFsid) {
+        let params = { get_alerts_from_fsid: getAlertsFromFsid };
+        this.prometheusService.getAlerts(params).subscribe(
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts:33]:
```suggestion
        this.prometheusService.getAlerts(ClusterFilteredAlerts).subscribe(
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts [hunk 6]
```diff
@@ -23,19 +23,31 @@ export class PrometheusAlertService {
...
-      );
+        );
+      } else {
+        this.prometheusService.getAlerts().subscribe(
+          (alerts) => this.handleAlerts(alerts),
+          (resp) => {
+            if ([404, 504].includes(resp.status)) {
+              this.prometheusService.disableAlertmanagerConfig();
+            }
+          }
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts:48]:
```suggestion

```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.ts [hunk 7]
```diff
@@ -43,7 +42,7 @@ export class DashboardV3Component extends PrometheusListHelper implements OnInit
   capacityService: any;
   capacity: any;
   healthData$: Observable<Object>;
-  prometheusAlerts$: Observable<AlertmanagerAlert[]>;
+  prometheusAlerts$: Observable<any>;
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.ts:45]: ?

**Review by nizamial09 — COMMENTED** [review]
looks good

**Comment by afreen23** [comment]
Tested with two ceph clusters - one quincy and other from latest:main

* I was unable to connect to quincy cluster, but the message that came up is not what we expect, perhaps failed earlier.
See the screencast.(ignore last part of dev console opening)
* Also it did not ask for fsid , did we remove it ?
* Moreover, I tried to connect another cluster (using latest:main) and it refused for that one too. Perhaps , I need to do some additional steps, let me know in that case.

PS: It was a fresh install using this PR, so everything was updated and loaded well which I can confirm!

[Screencast from 2024-04-06 00-02-26.webm](https://github.com/ceph/ceph/assets/25664409/f665eb6c-7d1d-4ec8-a21d-96635cd92426)

For alerts I can see cluster filter is set true always:

Screenshot from 2024-04-05 20-17-44

I wanted to test more after connecting a cluster but due to above could not do it


Thanks

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts [hunk 8]
```diff
@@ -54,8 +54,8 @@ export class PrometheusAlertService {
     });
   }
 
-  refresh() {
-    this.getAlerts();
+  refresh(clusterFilteredAlerts?: boolean) {
+    this.getAlerts(clusterFilteredAlerts);
     this.getRules();
```
[the remark this answers is not in the document]
Comment (cloudbehl) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts:59]: @aaSharma14

[the remark this answers is not in the document]
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts:59]: Rules don't have this issue as they are static and do not contain anything related to the fsid variable

**Comment by aaSharma14** [comment]
jenkins retest this please

**Comment by cloudbehl** [comment]
Also, tests are failing still.

**Review by nizamial09 — APPROVED** [review]

**Comment by afreen23** [comment]
When I connected a cluster not having multicluster, I was able to see the expected message
Screenshot from 2024-04-25 01-06-40
i did not see any error as mentioned above when connected both cluster supporting mult-cluster.
image

**Pull request merged**